### PR TITLE
pdf-sign: init at unstable-2022-08-08

### LINKED ISSUE #247987: Package request: pdf-sign
**Project description**

From the project page:

A tool to sign PDF files with Linux support.

*Why*

There appears to be a lack of applications that run on Linux and allow for attaching free-hand signatures to PDF files in a good way.

(Non-)alternatives include:

Xournal and Xournal++: Allows free-hand annotations but not saving those for later use. Allows inserting images and .pdf files but will rasterize them rather than retain the vector graphics, resulting in lower quality.
LibreOffice Draw: Allows inserting images and .pdf files but will rasterize them rather than retain the vector graphics, resulting in lower quality.
Okular: Allows free-hand annotations but not saving those for later use. Allows inserting custom saved stamps, but these will not be visible in other PDF readers.
Evince: Allows certain kinds of annotations, but nothing that can look like a free-hand signature.
pdftk, Ghostscript, and other command-line tools: Allows superimposing .pdf files, but not interactively selecting where.

**Metadata**

* homepage URL: https://github.com/svenssonaxel/pdf-sign
* source URL: https://github.com/svenssonaxel/pdf-sign/blob/master/pdf-sign
* license: MIT
* platforms: linux

### PR DESCRIPTION
**Pull request by TomaSajt**
## Description of changes

This PR adds 1 package: pdf-sign

Closes https://github.com/NixOS/nixpkgs/issues/247987

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by TomaSajt** [comment]
If anyone has experience using this program, please test out whether it is working as expected or not.

I managed to sign a pdf with a signature, however I couldn't figure out why I couldn't move the signature left or right in GUI (non-batch) mode.

File: pkgs/tools/graphics/pdf-sign/default.nix
Comment (drupol) [inline pkgs/tools/graphics/pdf-sign/default.nix:41]: If you are overriding `configurePhase`, `buildPhase`, `checkPhase`, `installPhase` or any other phase, you should not forget about explicitly running pre and post-phase hooks like [their](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L953) [original](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1008) [definitions](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1037) [do](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1078).

It is generally expected that an appropriate pre-phase hook (e.g. `preBuild`) hook will run at the beginning of a phase (e.g. `buildPhase`) and post-phase hook (e.g. `postBuild`) will run at the end. Hooks are normally ran as a part of a phase so if you override a phase as a whole, you will need to add `runHook hookName` calls (e.g. `runHook preBuild`) manually.

Having phases run pre/post-phase hooks is important because many setup hooks insert their own code into them – omitting a hook might therefore prevent some setup hooks required for proper functionality of a package from running. Additionally, hooks are often inserted by developers into the package expression and by users when overriding a package using `overrideAttrs`. Not running them can thus cause confusion why their code is not executed.

## Examples

### Before

```nix
  installPhase = ''
    your commands
  '';
```

### After

```nix
  installPhase = ''
    runHook preInstall

    your commands

    runHook postInstall
  '';
```

Source: https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/missing-phase-hooks.md

File: pkgs/tools/graphics/pdf-sign/default.nix
Comment (drupol) [inline pkgs/tools/graphics/pdf-sign/default.nix:47]: Please, keep as many attributes as possible sorted alphabetically, here the `meta` attributes needs to be sorted.

File: pkgs/tools/graphics/pdf-sign/default.nix
Comment (drupol) [inline pkgs/tools/graphics/pdf-sign/default.nix:30]: Can't we just add `ghostscript`, `pdftk`, `poppler_utils`, `coreutils` in the `buildInputs` so we can get rid of this patch?

File: pkgs/tools/graphics/pdf-sign/default.nix
Comment (drupol) [inline pkgs/tools/graphics/pdf-sign/default.nix:12]: `makeWrapper` doesn't seem to be used.

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I've left some comments, let me know if this is clear enough.

File: pkgs/tools/graphics/pdf-sign/default.nix
Comment (TomaSajt) [reply pkgs/tools/graphics/pdf-sign/default.nix:41]: Sorry, I always forget to do this :sweat_smile:

File: pkgs/tools/graphics/pdf-sign/default.nix
Comment (TomaSajt) [reply pkgs/tools/graphics/pdf-sign/default.nix:30]: I don't know how that would work. I don't think just having dependencies listed in `buildInputs` would allow the python script to call them, as this is only a shebang script, no special builders were used.

File: pkgs/tools/graphics/pdf-sign/default.nix
Comment (TomaSajt) [reply pkgs/tools/graphics/pdf-sign/default.nix:47]: I try to keep my attributes in a logical order, which might not be alphabetical.
I could put `description` to be first, sure, but having `maintainers` not be last is a bit weird IMO

Comment (drupol) [reply pkgs/tools/graphics/pdf-sign/default.nix:49]: Mmh I can understand. Since there is no strict rule on this, we can be flexible. I see many people ordering them naturally, so that's the reason why I'm requesting contributors to do so.

I won't complain if you don't want to do it, but for the sake of consistency and reproducibility, I will ask you to think twice and do it 😅

File: pkgs/tools/graphics/pdf-sign/default.nix
Comment (drupol) [reply pkgs/tools/graphics/pdf-sign/default.nix:28]: Have you tried ?

File: pkgs/tools/graphics/pdf-sign/default.nix
Comment (TomaSajt) [reply pkgs/tools/graphics/pdf-sign/default.nix:49]: Alright

File: pkgs/tools/graphics/pdf-sign/default.nix
Comment (drupol) [reply pkgs/tools/graphics/pdf-sign/default.nix:28]: Fair enough, let's keep it like that for the moment.

**Review by drupol — APPROVED** [review]
LGTM ! Thanks for your patience and for implementing all the requested changes.

Waiting for the CI to be done and merging afterward.

**Pull request merged**
